keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,345 @@
+
+### Story 2 : Database
+
+- Databases(Oracle, PostgreSQL, MySQL, MongoDB, SAP) consumes massive Memory and Hugepages, in order to reduce memory access latency and improve performance dramatically resources(CPU, Memory, Hugepages, I/O Devices) should be aligned.
+
+# Proposal
+
+## Proposed Changes
+
+### New Component: Memory Manager
+
+Memory Manager is a new component of Kubelet, it guarantees NUMA aware memory usage for the Pod which has Guarantedd QoS, here memory includes hugepages.
+
+To guarantee NUMA aware memory usage, when Guaranteed QoS Pod admission is requested, Memory Manager caculates NUMA node affinity to figure out appropriate NUMA node where enough memory and hugepages exist for containers in the Pod. Then, when Pod deployed, it enfoces containers in the Pod to consume memory and hugepages from designated NUMA node which Topology Manager selected.
+
+NOTE:
+- Memory Manager isolates Memory and Hugepages to single NUMA node to prevent remote(Inter-NUMA) memory access. For this, Memory Manager enforce cgroups cpuset subsystem using CRI API. For more details, see [CPUSET Enforcement](#cpuset-enforcement).
+
+- Memory Manager guarantees that the container(in Guanteed QoS Pod) consumes requested amount of memory(which is isolated to a NUMA node) through linux OOM system. For more details, see [OOM Score Enforcement](#oom-score-enforcement)
+
+More details of this component are listed below.
+
+#### The Concept of Safety Zone
+
+Memory Manager has the concept of safety zone to manage memory and hugepages at the NUMA node level. Basically, Kubernetes support reserving memory at the node level through Node Allocatable feature. It provides options such as `system-reserved` and `kube-reserved` to reserve CPUs, Memory, Storage for system services and kubernetes system components. Fortunately, it is possible to reserve CPUs with consideration of NUMA because CPUs are reserved using their logical core number. On the other hand, it is impossible to reserve memory with consideration of NUMA, the only thing we can do is reserving memory at the node level.
+
+The concept of safety zone is proposed to reserve certain amount of memory and hugepages equally on each NUMA node. The value of safety zone can be confiugred by configuration parameter of Memory Manager. For more details see the [Feature Gate and Kubelet Flags](#feature-gate-and-kubelet-flags) section.
```

Comment:
How do you propose to handle reserving hugepages of different sizes?

Reply:
At the impl level, the Memory Manager will have 2+ memory map:
1) one for conventional memory.
2) particular memory map for particular hugepage size.

Then when the NUMA node, where a container will be pinned, determined, cpuset.mems will be used to set NUMA restriction.

Reply:
I will mention it on KEP.

Reply:
I was wondering what the actual Kubelet flags and/or commandline args would look like.  `system-reserved` and `kube-reserved`, for example.

Reply:
Oh, i got it now, it might be part of below section.
I will show you example tomorrow.
- Proposed Configuration Flag:  
  `TBD`